### ISSUE TITLE
Compile order issue

### ISSUE THREAD

**Issue by Sachonski** [ValveSoftware/steamvr_unity_plugin#329]
We are trying to implement a custom library (not dll), and integrate it with the SteamVR project solution. But found out by updating it, that we cannot use that implementation anymore even tho the library in question is compiled before the SteamVR, SteamVR_Actions,SteamVR_Editor,SteamVR_Input_Editor and SteamVR_Windows_EditorHelper.
Still it's not letting us implement scripts from a pre compiled library.

Any idea how to fix this issue?

Thanks

**Comment by zite** [comment]
I'm not positive what the right fix is for this, I'll ask unity. In the meantime you can delete the assembly definition files from your project: 

/Assets/SteamVR/SteamVR.asmdef
/Assets/SteamVR/Editor/SteamVR_Editor.asmdef
/Assets/SteamVR/Input/Editor/SteamVR_Input_Editor.asmdef
/Assets/SteamVR/WindowsHelper/SteamVR_Windows_EditorHelper.asmdef
/Assets/SteamVR_Input/SteamVR_Actions.asmdef

**Comment by zite** [comment]
Can you describe more about the custom library you're trying to compile? If you also have an assembly definition file you may need to include the steamvr ones as dependencies.

**Comment by Sachonski** [comment]
We've tried deleting everything a few times and it did not worked.
It does not have a assembly definition since its a series of regular scripts that compile with the first pass (regular project scripts with namespaces).
It used to work like a charm on previous versions of SteamVR, but since now they are compiled in different stages in the new version, no longer works.
The solution was to implement new custom scripts that have the same functionality as the SteamVR ones. 
Its a bit of a dirty fix, but our game needed to be updated since SteamVR 2.15 was broken.

**Comment by zite** [comment]
I'm not sure I understand the issue. Can you create a small project to reproduce the error you're seeing?

**Comment by Sachonski** [comment]
Try using any SteamVR scripts, such as InteractableHoverEvents with a custom method that implements any script outside the SteamVR folder.

**Comment by zite** [comment]
This seems to work fine for me? Are you including the namespace in a using statement? I think I do what you're saying in the tutorial here: https://valvesoftware.github.io/steamvr_unity_plugin/tutorials/SteamVR-Input.html

Or in this video version here: https://youtu.be/qo-9CmcKWlY?t=917

**Comment by zite** [comment]
Oh, I think I see. Are you saying you're trying to modify a script included with the steamvr plugin? I don't think that's a supported scenario. You'd either need to create your own assembly definition file or delete the assembly definition files from the locations I mentioned. You may also need to delete the .sln and .csproj files in the root of your project (they'll get regenerated when selecting Open C# Project from the menu).

Alternatively if you're modifying SteamVR scripts I'd recommend creating your own versions of them since otherwise they'll get overwritten as you upgrade.

**Comment by Sachonski** [comment]
The thing is that it used to work in the old SteamVR 2.15 or bellow because the compile order of SteamVR was in the same as the project root script.
Yeah, actually we've made a custom hover similar to the SteamVR one and now everything works like a charm. 
Regarding the assembly definition, since they are compiled in the first pass, according to Unity everything that is compiled in previous steps of the compile order should be available for same or further step along the chain. 
(The .sln and .csrpoj was deleted multiple times and the project reimported several times.)

Anyhow, we've managed to solve the issue with a workaround, but in case it happens to anybody else now they've got a reference.
